library/mps_reader.c

```diff
@@ -0,0 +1,557 @@
+
+        /* Remember overlap of accumulator and fragment. */
+        rd->acc_share.frag_offset = aa;
+        aa += copy_to_acc;
+        rd->acc_available = aa;
+    }
+    else /* Not accumulating */
+    {
+        rd->acc_share.frag_offset = 0;
+    }
+
+    rd->frag = new_frag;
+    rd->frag_len = new_frag_len;
+    rd->commit = 0;
+    rd->end = 0;
+    MBEDTLS_MPS_TRACE_RETURN( 0 );
+}
+
+
+int mbedtls_mps_reader_get( mbedtls_mps_reader *rd,
+                            mbedtls_mps_size_t desired,
+                            unsigned char **buffer,
+                            mbedtls_mps_size_t *buflen )
+{
+    unsigned char *frag;
+    mbedtls_mps_size_t fl, fo, end, frag_fetched, frag_remaining;
+    MBEDTLS_MPS_TRACE_INIT( "mbedtls_mps_reader_get" );
+    MBEDTLS_MPS_TRACE( mbedtls_mps_trace_comment,
+                       "* Bytes requested: %u", (unsigned) desired );
+
+    MBEDTLS_MPS_STATE_VALIDATE_RAW( mps_reader_is_consuming( rd ),
+          "mbedtls_mps_reader_get() requires reader to be in consuming mode" );
+
+    end = rd->end;
+    fo = mps_reader_get_fragment_offset( rd );
+
+    /* Check if we're still serving from the accumulator. */
+    if( mps_reader_serving_from_accumulator( rd ) )
+    {
+        unsigned char *acc;
+
+        MBEDTLS_MPS_TRACE( mbedtls_mps_trace_comment,
+                           "Serve the request from the accumulator" );
+        if( fo - end < desired )
+        {
+            /* Illustration of supported and unsupported cases:
+             *
+             * - Allowed #1
+             *
+             *                          +-----------------------------------+
+             *                          |               frag                |
+             *                          +-----------------------------------+
+             *
+             *             end end+desired
+             *              |       |
+             *        +-----v-------v-------------+
+             *        |          acc              |
+             *        +---------------------------+
+             *                          |         |
+             *                   fo/frag_offset aa/acc_available
+             *
+             * - Allowed #2
+             *
+             *                          +-----------------------------------+
+             *                          |               frag                |
+             *                          +-----------------------------------+
+             *
+             *                  end          end+desired
+             *                   |                |
+             *        +----------v----------------v
+             *        |          acc              |
+             *        +---------------------------+
+             *                          |         |
+             *                   fo/frag_offset aa/acc_available
+             *
+             * - Not allowed #1 (could be served, but we don't actually use it):
```

Comment:
I prefer to leave it as is for now, because it's a means for us to double-check our understanding of the way the upper layers of MPS are using the readers.